# Importing libraries

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

# Importing dataframe

In [2]:
# Folder path
path = '/Users/katherinevaldivia/Desktop/Career Foundry Data Analytics Course work/Data Analytics Immersion/Achievement 4/06-23-Instacart Basket Analysis /'
path

'/Users/katherinevaldivia/Desktop/Career Foundry Data Analytics Course work/Data Analytics Immersion/Achievement 4/06-23-Instacart Basket Analysis /'

In [3]:
# Importing 'orders_products_merged_updated.pkl'

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))


# Following along with exercise 4.8

In [14]:
# Checking shape of dataframe
ords_prods_merge.shape

(32404859, 18)

In [15]:
# Checking first 5 of dataframe
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y product_name  aisle_id  department_id  prices _merge  \
0           195         Soda        77              7     9.0   both   
1           195         Soda        77              7     9.0   both   
2           195         Soda        77              7     9.0   both   
3           195         Soda        77              7     9.0   both   
4           195         Soda        77              7     9.0   both   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly days        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

In [8]:
# Creating subset
df = ords_prods_merge[:1000000]

In [16]:
# Checking shape of subset
df.shape

(1000000, 18)

In [17]:
# Checking first 10 rows in subset
df.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   
5   3367565        1             6                  2                  7   
6    550135        1             7                  1                  9   
7   3108588        1             8                  1                 14   
8   2295261        1             9                  1                 16   
9   2550362        1            10                  4                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   
5                   19.0         196                  1          1   
6                   20.0         196                  1          1   
7                   14.0         196                  2          1   
8                    0.0         196                  4          1   
9                   30.0         196                  1          1   

   Unnamed: 0_y product_name  aisle_id  department_id  prices _merge  \
0           195         Soda        77              7     9.0   both   
1           195         Soda        77              7     9.0   both   
2           195         Soda        77              7     9.0   both   
3           195         Soda        77              7     9.0   both   
4           195         Soda        77              7     9.0   both   
5           195         Soda        77              7     9.0   both   
6           195         Soda        77              7     9.0   both   
7           195         Soda        77              7     9.0   both   
8           195         Soda        77              7     9.0   both   
9           195         Soda        77              7     9.0   both   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly days        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders  
5  Regularly busy  Regularly days        Average orders  
6  Regularly busy    Busiest days           Most orders  
7  Regularly busy    Busiest days           Most orders  
8  Regularly busy    Busiest days           Most orders  
9      Least busy    Slowest days        Average orders

# Grouping Data with pandas

In [18]:
df.groupby('product_name')

In [20]:
#You should always use the groupby() function as part of a series of steps, namely, the following:

#1. Split the data into groups based on some criteria.
#2. Apply a function to each group separately.
#3. Combine the results into a dataframe or alternative data structure or create a new column in the current dataframe.

#You’ve only completed the first step—splitting the data into groups (with the group being the “product_name” column). 
#Now, let’s take a look into the second step, which will involve some aggregation!

# Aggregating Data with agg()

## Performing a Single Aggregation

In [21]:
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [22]:
df.groupby('department_id')['order_number'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

## Performing Multiple Aggregations

In [23]:
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

# Aggregating Data with transform()

## Creating a loyalty flag using loc

In [24]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [25]:
ords_prods_merge.head(15)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0    2539329        1             1                  2                  8   
1    2398795        1             2                  3                  7   
2     473747        1             3                  3                 12   
3    2254736        1             4                  4                  7   
4     431534        1             5                  4                 15   
5    3367565        1             6                  2                  7   
6     550135        1             7                  1                  9   
7    3108588        1             8                  1                 14   
8    2295261        1             9                  1                 16   
9    2550362        1            10                  4                  8   
10   2968173       15            15                  1                  9   
11   1870022       15            17                  2                 16   
12   1911383       15            18                  2                 11   
13   2715276       15            21                  1                  9   
14    487368       15            22                  1                 10   

    days_since_last_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   
5                    19.0         196                  1          1   
6                    20.0         196                  1          1   
7                    14.0         196                  2          1   
8                     0.0         196                  4          1   
9                    30.0         196                  1          1   
10                    7.0         196                  2          0   
11                    8.0         196                  6          1   
12                    7.0         196                  1          1   
13                    7.0         196                  1          1   
14                   14.0         196                  1          1   

    Unnamed: 0_y product_name  aisle_id  department_id  prices _merge  \
0            195         Soda        77              7     9.0   both   
1            195         Soda        77              7     9.0   both   
2            195         Soda        77              7     9.0   both   
3            195         Soda        77              7     9.0   both   
4            195         Soda        77              7     9.0   both   
5            195         Soda        77              7     9.0   both   
6            195         Soda        77              7     9.0   both   
7            195         Soda        77              7     9.0   both   
8            195         Soda        77              7     9.0   both   
9            195         Soda        77              7     9.0   both   
10           195         Soda        77              7     9.0   both   
11           195         Soda        77              7     9.0   both   
12           195         Soda        77              7     9.0   both   
13           195         Soda        77              7     9.0   both   
14           195         Soda        77              7     9.0   both   

       busiest_day    busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regularly days        Average orders         10  
1   Regularly busy    Slowest days        Average orders         10  
2   Regularly busy    Slowest days           Most orders         10  
3       Least busy    Slowest days        Average orders         10  
4       Least busy    Slowest days           Most orders         10  
5   Regularly busy  Regularly days        Ave

# Deriving Columns with loc()

In [63]:
# To create your flag, you’ll need some criteria. You can use the following:
# 1. If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
# 2. If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
# 3. If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [26]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [27]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'


In [28]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [29]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [30]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y product_name  aisle_id  department_id  prices _merge  \
0           195         Soda        77              7     9.0   both   
1           195         Soda        77              7     9.0   both   
2           195         Soda        77              7     9.0   both   
3           195         Soda        77              7     9.0   both   
4           195         Soda        77              7     9.0   both   

      busiest_day    busiest_days busiest_period_of_day  max_order  \
0  Regularly busy  Regularly days        Average orders         10   
1  Regularly busy    Slowest days        Average orders         10   
2  Regularly busy    Slowest days           Most orders         10   
3      Least busy    Slowest days        Average orders         10   
4      Least busy    Slowest days           Most orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [31]:
# check the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column

ords_prods_merge[['user_id', 'max_order', 'loyalty_flag']].sample(30)

user_id  max_order      loyalty_flag
10163671   117088         37  Regular customer
4011105     31403         17  Regular customer
1027951     20444          8      New customer
25182159    42216         34  Regular customer
22981626   178014          5      New customer
19359860   150673         25  Regular customer
3661734    194315         18  Regular customer
26654658   148057         18  Regular customer
10222195   158218         13  Regular customer
31335668   123519         14  Regular customer
21295753    11361         51    Loyal customer
14923325   135039         37  Regular customer
1199224    124481         52    Loyal customer
28514201   202068         38  Regular customer
29680136   167143         27  Regular customer
25192966    14144         21  Regular customer
25426624   146348         63    Loyal customer
1945722    182015         10      New customer
23316142    93532         16  Regular customer
5512325     31033          7      New customer
32213230    62495         25  Regular customer
20042610   165593         21  Regular customer
4376836     12132         15  Regular customer
22429717   161747         34  Regular customer
15820397    76412          9      New customer
23069633   114004          8      New customer
12175362    13527         70    Loyal customer
18769679    10613         66    Loyal customer
8381459    112015         20  Regular customer
14307344   185833         14  Regular customer

# 4.8 Tasks

## Step 1: Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise.

In [32]:
# Done.

## Step 2: find the aggregated mean for the entire dataframe.

In [33]:
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3: Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

In [34]:
# Results from subset
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

### 
Answer: The average results from the subset is greater than the average result of the entire dataframe except for department_id: 17

## Step 4: Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [35]:
# Done, Under deriving columns with loc section output 26.

## Step 5: The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [36]:
print('The mean, min and max of prices grouped by the loyalty_flag column in ords_prods_merge:')

ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in ords_prods_merge:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

### 
Answer: New customers hold the highest purchased price followed by regular customers and finally loyal customers. A quick thing to note is that the prices for max column seem too high.

## Step 6: Create a spending flag for each user based on the average price across all their orders using the following criteria:

### 1. If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.” 
### 2. If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [41]:
# Creating “average_price” column using transform()

ords_prods_merge['average_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

# Checking head top 5
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y  ... aisle_id  department_id  prices  _merge     busiest_day  \
0           195  ...       77              7     9.0    both  Regularly busy   
1           195  ...       77              7     9.0    both  Regularly busy   
2           195  ...       77              7     9.0    both  Regularly busy   
3           195  ...       77              7     9.0    both      Least busy   
4           195  ...       77              7     9.0    both      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag average_price  
0  Regularly days        Average orders        10  New customer      6.367797  
1    Slowest days        Average orders        10  New customer      6.367797  
2    Slowest days           Most orders        10  New customer      6.367797  
3    Slowest days        Average orders        10  New customer      6.367797  
4    Slowest days           Most orders        10  New customer      6.367797  

[5 rows x 21 columns]

In [42]:
# Rounding averages 

ords_prods_merge['average_price'] = ords_prods_merge['average_price'].round(2)

#Checking head top 5
ords_prods_merge.head()


order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y  ... aisle_id  department_id  prices  _merge     busiest_day  \
0           195  ...       77              7     9.0    both  Regularly busy   
1           195  ...       77              7     9.0    both  Regularly busy   
2           195  ...       77              7     9.0    both  Regularly busy   
3           195  ...       77              7     9.0    both      Least busy   
4           195  ...       77              7     9.0    both      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag average_price  
0  Regularly days        Average orders        10  New customer          6.37  
1    Slowest days        Average orders        10  New customer          6.37  
2    Slowest days           Most orders        10  New customer          6.37  
3    Slowest days        Average orders        10  New customer          6.37  
4    Slowest days           Most orders        10  New customer          6.37  

[5 rows x 21 columns]

In [43]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [44]:
# Creating a 'spending_flag' flag

ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [46]:
#Checking head top 5
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y  ... department_id  prices  _merge     busiest_day  \
0           195  ...             7     9.0    both  Regularly busy   
1           195  ...             7     9.0    both  Regularly busy   
2           195  ...             7     9.0    both  Regularly busy   
3           195  ...             7     9.0    both      Least busy   
4           195  ...             7     9.0    both      Least busy   

     busiest_days busiest_period_of_day max_order  loyalty_flag  \
0  Regularly days        Average orders        10  New customer   
1    Slowest days        Average orders        10  New customer   
2    Slowest days           Most orders        10  New customer   
3    Slowest days        Average orders        10  New customer   
4    Slowest days           Most orders        10  New customer   

   average_price spending_flag  
0           6.37   Low spender  
1           6.37   Low spender  
2           6.37   Low spender  
3           6.37   Low spender  
4           6.37   Low spender  

[5 rows x 22 columns]

In [47]:
#Checking frequency

ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31769965
High spender      634894
Name: spending_flag, dtype: int64

# Step 7:  

## Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
1. If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
2. If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
3. If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [51]:
# Creating “median_prior_orders” column using transform()

ords_prods_merge['median_last_orders'] = ords_prods_merge.groupby(['user_id'])['days_since_last_order'].transform(np.median)

In [52]:
# Checking head top 5
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y  ... prices  _merge     busiest_day    busiest_days  \
0           195  ...    9.0    both  Regularly busy  Regularly days   
1           195  ...    9.0    both  Regularly busy    Slowest days   
2           195  ...    9.0    both  Regularly busy    Slowest days   
3           195  ...    9.0    both      Least busy    Slowest days   
4           195  ...    9.0    both      Least busy    Slowest days   

  busiest_period_of_day max_order  loyalty_flag average_price  spending_flag  \
0        Average orders        10  New customer          6.37    Low spender   
1        Average orders        10  New customer          6.37    Low spender   
2           Most orders        10  New customer          6.37    Low spender   
3        Average orders        10  New customer          6.37    Low spender   
4           Most orders        10  New customer          6.37    Low spender   

  median_last_orders  
0               20.5  
1               20.5  
2               20.5  
3               20.5  
4               20.5  

[5 rows x 23 columns]

In [53]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_last_orders'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [54]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[(ords_prods_merge['median_last_orders'] > 10) & (ords_prods_merge['median_last_orders'] <= 20), 'order_frequency_flag'] = 'Regular customer'


In [55]:
# Creating a 'order_frequency_flag' flag

ords_prods_merge.loc[ords_prods_merge['median_last_orders'] <= 10, 'order_frequency_flag'] = 'Frequent customer'


In [56]:
ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2398795        1             2                  3                  7   
2    473747        1             3                  3                 12   
3   2254736        1             4                  4                  7   
4    431534        1             5                  4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

   Unnamed: 0_y  ... _merge     busiest_day    busiest_days  \
0           195  ...   both  Regularly busy  Regularly days   
1           195  ...   both  Regularly busy    Slowest days   
2           195  ...   both  Regularly busy    Slowest days   
3           195  ...   both      Least busy    Slowest days   
4           195  ...   both      Least busy    Slowest days   

   busiest_period_of_day max_order  loyalty_flag average_price spending_flag  \
0         Average orders        10  New customer          6.37   Low spender   
1         Average orders        10  New customer          6.37   Low spender   
2            Most orders        10  New customer          6.37   Low spender   
3         Average orders        10  New customer          6.37   Low spender   
4            Most orders        10  New customer          6.37   Low spender   

   median_last_orders   order_frequency_flag  
0                20.5  Non-frequent customer  
1                20.5  Non-frequent customer  
2                20.5  Non-frequent customer  
3                20.5  Non-frequent customer  
4                20.5  Non-frequent customer  

[5 rows x 24 columns]

In [57]:
# Checking frequency

ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [58]:
# There are 5 NaN values so I'll be checking them out.
ords_prods_merge[ords_prods_merge['order_frequency_flag'].isnull()][['user_id', 'days_since_last_order', 'median_last_orders', 'order_frequency_flag']]

user_id  days_since_last_order  median_last_orders  \
13645692   159838                    NaN                 NaN   
17251990   159838                    NaN                 NaN   
17622767   159838                    NaN                 NaN   
24138593   159838                    NaN                 NaN   
25880002   159838                    NaN                 NaN   

         order_frequency_flag  
13645692                  NaN  
17251990                  NaN  
17622767                  NaN  
24138593                  NaN  
25880002                  NaN

### All 5 NaN values come from customer user_id : 159838. The customers "days_since_last_order", "median_last_orders", and "order_frequency_flag" is NaN. This seems to indicate that this is a new customer. So there is no need to make any changes.

In [60]:
ords_prods_merge.shape

(32404859, 24)

# Step 8: Ensure your notebook is clean and structured and that your code is well commented.

In [61]:
# Check.

# Step 9: Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [62]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_updated_2.pkl'))